Reader: refactor infinite-stream component

samouri opened the pull request:
This PR seeks to create a new `InfiniteLoader` esque component implemented with react-virtualized.  That way we can reuse it throughout reader code.

This PR creates the new component and replaces `SitesWindowScroller` with it.
It will also be used in https://github.com/Automattic/wp-calypso/pull/13198

Things I can see us doing with this component in the future:
- ~~adding metrics~~ done
- adding arrow key support through https://github.com/bvaughn/react-virtualized/blob/master/docs/ArrowKeyStepper.md
- adding the ability to render posts as well so that it could be used in tandem with the reduxified streams.  it should ideally be as simple as adding a new row-renderer

TODO
- [x] fix loading bug
- [x] create the composable version of rowRenderer

To test:
1. load this branch 
2. go to /following/manage and make sure all of the lists there work as expected. that includes searching followed, scrolling through followed, searching feeds, and clicking show more to see more feeds.
3. go to /following/search and make sure that site results still render properly (should also be more performant)
3. railcars should also be fired for following-manage

samouri commented on client/components/reader-infinite-stream/row-renderers.js:
```diff
@@ -0,0 +1,26 @@
+ */
+import ConnectedSubscriptionListItem from 'blocks/reader-subscription-list-item/connected';
+
+export const siteRowRenderer = ( {
```
this is a bit weird still. should be items instead of sites and infinite-stream shouldn't need to know about followSource etc.

blowery commented:
hmmm seeing http://calypso.localhost:3000/following/manage?q=powazek&showMoreResults=true blow up looking for a missing rowRenderer?

blowery commented:
think i fixed the issues i was seeing, but could use some eyes.

samouri commented:
just realized its missing traintracks -- i'll add those now

samouri commented:
We hadn't been using traintracks before for the lists, only for the 10 before showMore.  This is now creating a ton of network requests

samouri commented:
something definitely broke from a perf perspective when i added traintracks -- investigating now

blowery commented:
> something definitely broke from a perf perspective when i added traintracks

looks like you broke component purity by passing a new object to the row renderer on each render pass... or did you mean a different kind of perf?

samouri commented:
> looks like you broke component purity by passing a new object to the row renderer on each render pass... or did you mean a different kind of perf?
 
yep :(.  I'll have a fix up shortly

samouri commented:
fixed the perf issue w/ traintracks

Two more issues:
- [ ] as you scroll through the results and they are still loading your scrollY bounces you about
- [ ] need to pay attention to search results w/ less than 200 result again

samouri commented:
rebased

samouri commented:
the strange scrollY issue seems to be 99% fixed.  I'm not sure how.

samouri commented:
I have no idea why I added forceRefresh, it was incredibly dumb. Clearing rowHeight cache is necessary whenever someone resizes the window because we may need to recalc heights for all the items, but it is totally unnecessary for reordering lists/ adding new items to the end!

The latest couple patches:
1. move more logic from following-manage to `subscriptions.jsx`. Including how much more to page.  It should now handle when we know es only has N < 200 results.
2. removed forceRefresh and instead have something called pasthroughProp.  `passthroughProp` just gets passed through to the underlying `List`

samouri commented on client/blocks/reader-subscription-list-item/connected.jsx:
```diff
@@ -38,7 +38,6 @@ class ConnectedSubscriptionListItem extends React.Component {
 	};
 
 	componentDidMount() {
-		this.props.onLoad();
```
no point of measuring right when it mounts -- that happens anyway. What matters is to measure each time the props change

samouri commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
-		const feedId = +get( site, 'feed_ID' );
-		const siteId = +get( site, 'blog_ID' );
+	// todo what is eventName / ui_algo?
+	recordTraintrackForRowRender = ( { index, railcar, eventName } ) => {
```
the idea here is that we can record traintracks for any arbitrary at the time of render just given a railcar and an eventName (rv supplies the index)

samouri commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
-		const feedUrl = get( site, 'feed_URL' );
-		const feedId = +get( site, 'feed_ID' );
-		const siteId = +get( site, 'blog_ID' );
+	// todo what is eventName / ui_algo?
```
are eventName + ui_algo always the same?

blowery replied:
ui_algo is supposed to be the "ui algorithm" we used to display it. initial intent was to just use the page name.

samouri commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
+			);
+		}
+
+		return this.props.rowRenderer( {
```
In order to make this as generic as possible, i've modified InfiniteStream so that instead of containing a `siteRowRenderer`, it actually receives a rowRenderer as a prop.  

Similar to how InfiniteList takes a `renderItem` func

samouri commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
+		>
+			{ ( { measure } ) => (
+				<div key={ key } style={ style } className="reader-infinite-stream__row-wrapper">
+					<ComponentToMeasure { ...props } onLoad={ measure } />
```
potentially should rename `onLoad` to `onShouldRemeasure`

samouri commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -124,22 +131,18 @@ class ReaderInfiniteStream extends Component {
-									/>
-								) }
-							</AutoSizer>
+							<List
```
autosizer is overkill now that we have withWidth. it also didn't work

samouri commented on client/reader/following-manage/feed-search-results.jsx:
```diff
@@ -1,115 +1,115 @@
-	const classNames = classnames( 'following-manage__search-results', {
-		'is-empty': isEmpty,
-	} );
+class FollowingManageSearchFeedsResults extends React.Component {
```
whoops.  I refactored this file more than I should have.

blowery replied:
?

samouri replied:
Everything is good about it. Was able to consolidate a couple of logic branches.  It just shouldn't have been intermixed with this PR that is focused on swapping out SitesWindowScroller --> infinite-stream

blowery commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
+		extraRenderItemProps: PropTypes.object,
+		minHeight: PropTypes.number,
+		renderEventName: PropTypes.string,
+		passthroughProp: PropTypes.any,
```
what's passthroughProp?

samouri replied:
When we change underlying data we need a way to inform rv since we don't actually ever give rv the underlying list that is backing the rowRenderers.  E.g. one piece of data we always give rv is the rowCount -- if rowCount changes, then rv knows to rerender.  

What if the sort changes though? 
Its all best explained here: https://github.com/bvaughn/react-virtualized#pure-components

Our two options are:
1. create a single passthroughProp 
2. create a a passthroughProps object that gets expanded into the list via `{ ..passthroughProps }`

blowery commented:
getting failures in sites in search

<img width="751" alt="screenshot 2017-06-07 16 48 34" src="https://user-images.githubusercontent.com/14350/26900743-3622ea0a-4ba1-11e7-904c-d1203e4a3e93.png">

blowery commented:
hmmm... not seeing `calypso_traintracks_interact` events fire when following a site from site search on following manage. also `calypso_traintracks_render` refires for all rendered results when i follow a site.

blowery commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,80 @@ import {
-		);
+	rowRenderer = rowRendererProps => {
+		const railcar = get( this.props.items[ rowRendererProps.index ], 'railcar' );
+		if ( railcar && this.props.renderEventName ) {
```
seems like this is firing too often. should only fire when rendering the first time.

samouri commented:
> hmmm... not seeing calypso_traintracks_interact events fire when following a site from site search on following manage

I think I need to fix this before merging

> also calypso_traintracks_render refires for all rendered results when i follow a site.
 
Since we never had render traintracks working properly, I think i'd rather strip it from this PR and solve that issue in a followup.  It may be a bit involved

samouri commented:
> Since we never had render traintracks working properly, I think i'd rather strip it from this PR and solve that issue in a followup. It may be a bit involved

ended up being really easy to fix, traintracks should be working in this pr now

blowery commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
-		const feedUrl = get( site, 'feed_URL' );
-		const feedId = +get( site, 'feed_ID' );
-		const siteId = +get( site, 'blog_ID' );
+	recordedRender = new Set(); // used to ensure we only fire render event once per item
```
if we move from one search to another, do we need to spin up a new set?

blowery replied:
looks like we're creating a new instance every time, so should be fine

blowery commented on client/components/reader-infinite-stream/index.js:
```diff
@@ -8,72 +8,81 @@ import {
+
+	rowRenderer = rowRendererProps => {
+		const railcar = get( this.props.items[ rowRendererProps.index ], 'railcar', undefined );
+		if ( railcar && this.props.renderEventName && ! this.recordedRender.has( rowRendererProps.index ) ) {
```
do we need to pass a renderEventName from `/read/search` for the site listing?

blowery commented on client/reader/search-stream/site-results.jsx:
```diff
@@ -29,18 +33,21 @@ class SitesResults extends React.Component {
 					fetchNextPage={ this.fetchNextPage }
-					forceRefresh={ searchResults }
+					hasNextPage={ this.hasNextPage }
+					rowRenderer={ siteRowRenderer }
```
do we need a `renderEventName` here?

samouri replied:
yes we do if we want the event to have an action! added
